Improve gen2 map importer

penev92 opened the pull request:
- Pull out a base type for the TS map importer, to be used as base for the RA2 map importer
- Fix map Bounds calculations (that I would love to rename to InteractableBounds and add VisibleBounds as well, since those differ in TS/RA2)
- Update gen1 map importer to match naming
- Reimport a bunch of TS maps

The 8 official maps can most easily be found in the CnCNet version of TS (or the original TS CD's mixes), and for the other 3 that I reimported I added a comment with a link to the original TS map file as well as uploaded here:
[custom_map_originals.zip](https://github.com/OpenRA/OpenRA/files/11176919/custom_map_originals.zip)

The next steps would be

- OpenRA:
  - Change cell (un)projection to match TS.
  - Add Map.VisibleBounds (alternatively a Shroud.VisibilityMargin) to account for TS's visible-but-not-interactable region(s)
- RA2:
  - Update the map importer on top of the new base importer.

penev92 committed:
Moved `ImportLegacyMapCommand` to OpenRA.Mods.Cnc

Also renamed to `ImportGen1MapCommand`.
Also moved Extensions.DistinctBy().

penev92 committed:
Added abstract base class ImportGen2MapCommand

Split out ImportGen2MapCommand from ImportTSMapCommand, to also serve as a base for a RA2 map importer.
Also renamed TS importer to ImportTiberianSunMapCommand.

pchote committed:
Fixed gen2 map importer Bounds calculations

penev92 committed:
Reimported some TS maps

Could not find the source for most of the maps, so reimported only those that were part of the original game's "official map pool" + 3 custom ones.

PunkPun commented on OpenRA.Mods.Cnc/UtilityCommands/ImportGen2MapCommand.cs:
```diff
@@ -683,5 +395,141 @@ static void ReadLamps(Map map, IniFile file)
+
+		protected MPos ToMPos(int dx, int dy)
+		{
+			// Convert TS relative position to OpenRA MPos, accounting for map cordons.
```
This should become the function summary

PunkPun commented on OpenRA.Mods.Cnc/UtilityCommands/ImportGen2MapCommand.cs:
```diff
@@ -356,21 +126,20 @@ static void ReadTiles(Map map, IniFile file, int2 fullSize)
 
 				var dx = rx - ry + fullSize.X - 1;
 				var dy = rx + ry - fullSize.X - 1;
```
this pattern is commonly repeated, perhaps add another ToMPos function with these inputs?

PunkPun commented:
To fix the mono error you also need to move Extensions.cs into CnC

Mailaender commented on OpenRA.Mods.Cnc/UtilityCommands/ImportGen1MapCommand.cs:
```diff
@@ -15,19 +15,20 @@
-namespace OpenRA.Mods.Common.UtilityCommands
+namespace OpenRA.Mods.Cnc.UtilityCommands
 {
-	public abstract class ImportLegacyMapCommand
```
I suggest keeping the name as we use the same for **Legacy**BridgeLayer/Hut and used it for the sprite sequences as well.

pchote replied:
The "Legacy" in LegacyBridgeLayer actually represents "this code is obsolete and will be removed as soon as someone ports RA and TD to use the new TS logic", which is different to its use here.

penev92 commented:
So I rebased just in case any new code style violations would pop up (they didn't). I reimported another map - The Way to Uganda, provided by Icesinithol on Discord (apparently a Yuri's Revenge map by Morton, converted to TS by Icesinithol).
I also added an extra commit showing what @PunkPun asked for, but I am not convinced that is better.

pchote commented on OpenRA.Mods.Cnc/UtilityCommands/ImportGen1MapCommand.cs:
```diff
@@ -452,4 +453,20 @@ void ReadTrees(IniFile file)
+	public static class Extensions
+	{
+		/// <summary>
+		/// Only used for Mono builds. .NET 6 added the exact same thing.
```
Do we want to wrap this in `#if !NET6_0_OR_GREATER` then?

pchote commented on OpenRA.Mods.Cnc/UtilityCommands/ImportGen2MapCommand.cs:
```diff
@@ -410,85 +163,28 @@ static void ReadOverlay(Map map, IniFile file, int2 fullSize)
-				// TS maps encode the non-harvestable border tiles as overlay
-				// Only convert to vein resources if the overlay data specifies non-border frames
-				if (overlayType == 0x7E)
+				if (TryHandleResourceFromOverlayInner(overlayType, out var resourceType))
```
IMO this should be returning the `ResourceTile` in case implementations want to modify the density.

pchote commented on OpenRA.Mods.Cnc/UtilityCommands/ImportGen2MapCommand.cs:
```diff
@@ -380,6 +396,29 @@ protected virtual void ReadLamps(Map map, IniFile file)
+			var playableAreaWidth = iniBounds[2];
+			var playableAreaHeight = 2 * iniBounds[3] - UninteractableMargin.Y;
+
+			var unknownHeightPadding = 0;
```
`var unknownHeightPadding = map.Height.Max() != 0 ? 8 : 0;`

pchote commented on OpenRA.Mods.Cnc/UtilityCommands/ImportGen2MapCommand.cs:
```diff
@@ -532,6 +527,13 @@ protected MPos ToMPos(int dx, int dy)
 			return new MPos(Cordon.Left + (dx + Cordon.Top % 2) / 2, Cordon.Top + dy);
 		}
 
+		protected MPos ToMPos(int rx, int ry, int mapWidthWithoutCordon)
```
I guess this is fine. Squash into the bounds commit?

pchote reviewed (commented):
Code changes look fine. Not yet tested ingame results.

penev92 commented:
I should also add the Juggernaut and Nod Artillery to `DeployableActors`.

penev92 commented:
Addressed the above comments.

PunkPun commented:
Rebase me!

penev92 commented:
Rebased.

PunkPun commented on OpenRA.Mods.Cnc/UtilityCommands/ImportGen2MapCommand.cs:
```diff
@@ -683,5 +387,148 @@ static void ReadLamps(Map map, IniFile file)
+			return new MPos(Cordon.Left + (dx + Cordon.Top % 2) / 2, Cordon.Top + dy);
+		}
+
+		protected MPos ToMPos(int rx, int ry, int mapWidthWithoutCordon)
```
these functions should share the summary

penev92 replied:
Done.

PunkPun reviewed (commented):
Otherwise LGTM. I haven't tried to import though

PunkPun commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/928952246db51a159097a42d9ece10dbd2defb59...34e69877c05051a446a33767be9fcbb12a6d6a2e)
